### DISCUSSION TITLE
3D visualization - building_elevation

### DISCUSSION THREAD

**Discussion by kmamor** [gama-platform/gama#338] [Q&A]
I am having difficulty getting my  model to display 3D. I am going back to basics and trying to get it to display in 2D. The following code works:

```
global {
	file shape_file_buildings_comm <- shape_file('../includes/aylesbury_buildings_comm.shp');
	file shape_file_buildings_res <- shape_file('../includes/aylesbury_buildings_res.shp');
	file shape_file_roads_all <- shape_file('../includes/aylesbury_roads_2.shp');
	
	geometry shape <- envelope(shape_file_roads_all);
	
	init {
		create building_res from: shape_file_buildings_res {
						}
	
		create building_comm from: shape_file_buildings_comm {
						}
		create road from: shape_file_roads_all;
			}
		}

species building_res {
rgb color <- #black;
aspect base {
	draw shape color: color ;
	// depth: height;
}	
	}

species building_comm {
rgb color <-#black;
aspect base {
	draw shape color: color;
	}	
	}
	
species road {
	rgb color <- #yellow;
	aspect base {
		draw shape color: color;
	}
}

experiment Tuto3D type: gui {
	
	output {
		display map  {
			species road aspect: base;
			species building_res aspect: base;
			species building_comm aspect: base;
				}
	}
	}

```

This properly displays a road network and buildings. A few queries. 

If I replace the geometry line with this:

`geometry shape <- envelope(rectangle(10 #miles, 5 #miles ));`

Then nothing displays. The dimensions reflect the dimensions of the underling road network (as measured in OSM where the shapefiles were created). 

Second query: in the building_elevation model a file 'bounds.shp' is included which sets the 3D environment. How would one make such a file given a set 2D road network and a desired height?

**Comment by ptaillandier** [comment]
Hi,

For the first question (using  "geometry shape <- envelope(rectangle(10
#miles, 5 #miles ))"), you get this result because GAMA is not able to
locate the environment in this case. In fact, "geometry shape <-
envelope(shape_file_roads_all)" not only sets the size of the world, but
also its coordinate reference system and allows to locate the environment
in the real world. If you are using shapefile, you need to use the envelope
of a geographic data file (or files) to set the shape of the world.

For question 2 about the height of the roads, I am not sure to understand.
Do you have data about the height of the different roads?

Cheers,

Patrick

**Reply by kmamor** [reply]
Thanks Patrick,

I understand.  Let's say that I wish to use the roads and buildings in a 3D model but with height in order that some agents can fly over the cityscape from one building to another.  Let's say a height of 1 km. 

How would I create a suitable shapefile to represent the  environment (this is why I asked the second question - how was the bounds.shp file created which presumably holds the location data and distances as well as height)?   The road and buildings shapefiles were created in QGIS using OSM data. 

Thanks,

Kevin

**Reply by kmamor** [reply]
In QGIS I created an new layer, covered the road network with a rectangle and then gave the rectangle height and saved the layer as a shapefile. This seems to have solved one problem. I now get the 3D environment to display. See attached images from top and front. 
![front_view](https://github.com/user-attachments/assets/76f7ea9d-a125-40d6-ba0d-be6ef3846831)
![top_view](https://github.com/user-attachments/assets/73ad1727-0f46-42b5-9c65-a5a5f60985b1)

With the following code I add drones (the orange spheres in the images). 

```
model aylesbury3

global {
	
	file shape_file_buildings_comm <- shape_file('../includes/aylesbury_buildings_comm.shp');
	file shape_file_buildings_res <- shape_file('../includes/aylesbury_buildings_res.shp');
	file shape_file_roads_all <- shape_file('../includes/aylesbury_roads_2.shp');
	file shape_file_bounds <- shape_file('../includes/aylesbury_environ.shp');
	int nb_drones <- 10;
	int environment_size <- 20;
	geometry shape <- envelope(shape_file_bounds);
	graph road_network;
	
	init {
		create building_res from: shape_file_buildings_res {
				height <- 10 + rnd(90);
				}
			create building_comm from: shape_file_buildings_comm {
				height <- 10 + rnd(90);
				}
		create road from: shape_file_roads_all;
		road_network <- as_edge_graph(road);
		create drone number: nb_drones {
			 location <- {rnd(1,18000), rnd(1,10000), rnd(1,1000)};	
				}  		} 	}

species drone skills: [moving3D] {
	rgb color;
	
	reflex move {
		do wander;
	}
	aspect base {
		draw sphere(environment_size * 10) color: #orange;
			}  }

species building_res {
rgb color <- #black;
int height ; 
aspect base {
	draw shape color: color depth: height;
}		}

species building_comm {
rgb color <-#black;
int height ; 
aspect base {
	draw shape color: color depth: height;
}		}
	
species road {
	rgb color <- #yellow;

	aspect base {
		draw shape color: color;
	} }

experiment Tuto3D type: gui {
	
	output {
		display map type: 3d {
			species road aspect: base;
			species building_res aspect: base;
			species building_comm aspect: base;
			species drone aspect: base;			
		}  	}  }

```


However the drone's move reflex isn't working - nothing happens when I run the experiment. Does anyone have suggestions as to why that might be? 

Kevin
